### DISCUSSION TITLE
libvips memory usage

### DISCUSSION THREAD

**Discussion by fcuenca4** [libvips/libvips#3338] [Q&A]
Hi! First of all, thanks for the great work you are putting into libvips.

I have a long running k8s service that is currently converting images from different formats to jpeg on a synchronous way.
The service is currently being hosted in EKS t2.large node with 6GB or ram and 1vCPU and, after doing some load testing, it can't process more than a hundred 300kb conversions without exhausting all the memory and getting OOM.

I'm using libvips 8.14.1 with concurrency=2 cache_max_files=0 cache_max_mem=52428800 cache_max=100.

I'm using the `copy` command with `Q=100`.

What's could be the reason for this behaviour? Is there any ideas that might help me debug the issue? I'm running out of :(

Some alternatives I've tried after reading multiple discussions: 
* using jemalloc 
* MALLOC_ARENA_MAX=2  
* openjpeg

Thanks
Fran

**Comment by jcupitt** [comment]
Hi @fcuenca4,

What image formats are you converting, and how are you converting them? Did you enable sequential mode? I made a tiny demo prog for you:

```python
#!/usr/bin/python3

import os
import sys
import glob

import pyvips

for filename in glob.glob(f"{sys.argv[1]}/*"):
    directory, file = os.path.split(filename)
    base, ext = os.path.splitext(file)
    print(f"processing {base} ...")
    x = pyvips.Image.new_from_file(filename, access="sequential")
    x.write_to_file(f"{sys.argv[2]}/{base}-q100.jpg", Q=100)
```

I can run it like this:

```
$ mkdir samples samples2
$ for i in {1..10000}; do cp ~/pics/shop.jpg samples/$i.jpg; done
$ ls -l ~/pics/shop.jpg
-rw------- 1 john john 395831 Feb 13  2021 /home/john/pics/shop.jpg
$ vipsheader ~/pics/shop.jpg
/home/john/pics/shop.jpg: 2364x1632 uchar, 3 bands, srgb, jpegload
$ /usr/bin/time -f %M:%e ./cvt.py samples samples2
...
processing 2233 ...
230320:490.65
$
```

So it converted 10,000 400kb images in 500s and had a peak memory use of 230kb.

**Reply by fcuenca4** [reply]
I've tried via different ways, first using govips and feeding the buffer
```
func (p *Photo) Convert() (*bytes.Reader, *vips.ImageMetadata, error) {
	rawBytes, metadata, err := p.image.ExportJpeg(&vips.JpegExportParams{
		StripMetadata: false,
		Quality:       100,
		Interlace:     false,
	})
	if err != nil {
		return nil, nil, err
	}

	return bytes.NewReader(rawBytes), metadata, nil
}
```

but I've also tried by installing vips-tools and using the command line
```
func (p *Photo) Convert() (*bytes.Reader, *vips.ImageMetadata, error) {
	start := time.Now()
	cmd := exec.Command("/usr/bin/vips", "copy", "stdin", ".jpg[Q=100]")
	cmd.Stdin = p.buffer
	logger := yule.NewSimpleLogger()
	out, err := cmd.CombinedOutput() 
	... more code
```
None of them worked, memory is still not being freed correctly :(
I've also tried with different image types but mainly HEIF files.
I'm currently reading from a memory buffer, is sequential still available?

**Reply by jcupitt** [reply]
Yes, eg.:

```python
#!/usr/bin/python3

import os
import sys
import glob

import pyvips

for filename in glob.glob(f"{sys.argv[1]}/*"):
    directory, file = os.path.split(filename)
    base, ext = os.path.splitext(file)
    print(f"processing {base} ...")

    with open(filename, "rb") as f:
        data = f.read()

    x = pyvips.Image.new_from_buffer(data, "", access="sequential")
    y = x.write_to_buffer(".jpg", Q=100)

    with open(f"{sys.argv[2]}/{base}-q100.jpg", "wb") as f:
        f.write(y)
```

I tried with 1,000 800kb HEIC files:

```
$ vipsheader sample/1.heic 
sample/1.heic: 3456x4608 uchar, 3 bands, srgb, heifload
$ /usr/bin/time -f %M:%e ./cvt.py sample sample2
processing 938 ...
...
processing 84 ...
2758224:537.83
```

So 2.5gb and 540s. Memory use is higher since HEIC decode needs a *lot* of memory, but it should be stable.

**Comment by kleisauke** [comment]
Cross-posted here: https://github.com/davidbyttow/govips/issues/341.

**Comment by jcupitt** [comment]
Ahh please don't cross-post questions, it makes extra unnecessary work for volunteer maintainers. Do I have to write two answers now? :( sigh.

**Reply by fcuenca4** [reply]
@jcupitt sorry for the inconvenience, I've closed #341
